sql: fix possible race in runPlanInsidePlan

yuzefovich opened the pull request:
This commit fixes a data race that was introduced in 187eabd0692fbfc9fe76affb5adf95a9eb49cf08. In particular, in that commit we made it so that `evalCtxFactory` method, in addition to creating a fresh copy of the eval context, also modifies the planner to point to the updated eval context. I believe this was needed so that `deferredRoutineSender` could be visible from the planner. However, `evalCtxFactory` method can be called concurrently (in the parallel checks code path), yet we'd be modifying the same `plannerCopy` object. This race is now eliminated by setting the deferred routine sender once on the eval context of the planner copy, meaning that the deferred routine sender will be visible from all code paths inside of `runPlanInsidePlan`.

The second commit in this PR is the reduced version of the commit that is reverted in the first commit.

Fixes: #116748.

Release note: None

yuzefovich committed:
Revert "sql: fix possible race in runPlanInsidePlan"

This reverts commit 0313990b8364c559612dbd2617ae70b4f79c6e6e.

That commit introduced an optimization of reusing the same
`serialEvalCtx` when the eval ctx factory is not used concurrently, but
this deviates from the main query path where we do deep reset of the
eval ctx between usages.

yuzefovich committed:
sql: fix possible race in runPlanInsidePlan

This commit fixes a data race that was introduced in 187eabd.
In particular, in that commit we made it so that `evalCtxFactory`
method, in addition to creating a fresh copy of the eval context, also
modifies the planner to point to the updated eval context. I believe
this was needed so that `deferredRoutineSender` could be visible from
the planner. However, `evalCtxFactory` method can be called concurrently
(in the parallel checks code path), yet we'd be modifying the same
`plannerCopy` object. This race is now eliminated by setting the
deferred routine sender once on the eval context of the planner copy,
meaning that the deferred routine sender will be visible from all code
paths inside of `runPlanInsidePlan`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116963)
<!-- Reviewable:end -->

yuzefovich commented:
I'll backport the second commit to 23.2. Also, I'll do some cleanup around the eval ctx methods in a separate PR because `ExtendedEvalContextReset` and `ExtendedEvalContextCopyAndReset` are confusing and don't do what their names suggest they do.

yuzefovich committed:
sql: remove confusing eval ctx methods

This commit is a partial revert of 54c3e26fd84cc4defe16129403008fc83856e435.
The main goal of this change is the removal of
`ExtendedEvalContextCopyAndReset` and `ExtendedEvalContextReset` because
they don't do what their names suggest they do.

I think it was unnecessary that 54c3e26fd84cc4defe16129403008fc83856e435
made any changes around this:
- that commit didn't end up using these two methods in
`runPlanInsidePlan`, so these two methods extracted functionality only
used in one place
- this functionality is not needed in the `runPlanInsidePlan` because we
don't start with a fresh `extendedEvalCtx` there - we always make
a copy of already initialized one (unlike in the main query path where
we start with a fresh copy and then initialize it and set all the
necessary fields).

Release note: None

yuzefovich commented:
Actually, I think it makes sense to include the cleanup in this PR, so I added third commit here.

yuzefovich commented:
I'll backport only the second commit to 23.2.x which fixes the data race.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116963#-:-NnA9ONw4qcghjoYpRUV:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rharding6373)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
